File: integrations/chatgpt/agenttrust_client.py
"""
AgentTrust Client for ChatGPT Integration
Provides a Python client for ChatGPT to interact with AgentTrust API

Auth0 for AI Agents Hackathon: Built with Token Vault for OAuth flows,
token management, and consent delegation.
"""

import requests
import json
import time
from datetime import datetime
from typing import Dict, Optional, Any
import os

# Load .env file if python-dotenv is available
try:
    from dotenv import load_dotenv
    _dir = os.path.dirname(os.path.abspath(__file__))
    _root = os.path.dirname(os.path.dirname(_dir))
    for p in [os.path.join(_root, 'backend', '.env'), os.path.join(_root, '.env'), os.path.join(_dir, '.env')]:
        if os.path.isfile(p):
            load_dotenv(p)
except ImportError:
    pass


class AgentTrustClient:
    """Client for interacting with AgentTrust API from ChatGPT"""
    
    def __init__(
        self,
        api_url: str = None,
        auth0_domain: str = None,
        auth0_client_id: str = None,
        auth0_client_secret: str = None,
        auth0_audience: str = None
    ):
        """
        Initialize AgentTrust client
        
        Args:
            api_url: AgentTrust API URL (default: from env or http://localhost:3000/api)
            auth0_domain: Auth0 domain
            auth0_client_id: Auth0 client ID
            auth0_client_secret: Auth0 client secret
            auth0_audience: Auth0 API audience
        """
        self.api_url = api_url or os.getenv('AGENTTRUST_API_URL', 'http://localhost:3000/api')
        self.auth0_domain = auth0_domain or os.getenv('AUTH0_DOMAIN')
        self.auth0_client_id = auth0_client_id or os.getenv('AUTH0_CLIENT_ID')
        self.auth0_client_secret = auth0_client_secret or os.getenv('AUTH0_CLIENT_SECRET')
        self.auth0_audience = auth0_audience or os.getenv('AUTH0_AUDIENCE')
        self.dev_mode = os.getenv('AGENTTRUST_DEV_MODE', 'false').lower() == 'true'
        
        self._token = None
        self._token_expiry = None
        self.current_session_id = None
        self.current_prompt_id = None
        
        if not self.dev_mode and not all([self.auth0_domain, self.auth0_client_id, 
                   self.auth0_client_secret, self.auth0_audience]):
            raise ValueError(
                "Auth0 credentials must be provided or set in environment. "
                "Or set AGENTTRUST_DEV_MODE=true to run without backend (browser only)."
            )
    
    def _get_token(self) -> str:
        """Get Auth0 access token (with caching)"""
        # Check if token is still valid
        if self._token and self._token_expiry and datetime.now() < self._token_expiry:
            return self._token
        
        # Request new token
        response = requests.post(
            f"https://{self.auth0_domain}/oauth/token",
            json={
                "client_id": self.auth0_client_id,
                "client_secret": self.auth0_client_secret,
                "audience": self.auth0_audience,
                "grant_type": "client_credentials"
            }
        )
        
        if response.status_code != 200:
            raise Exception(f"Failed to get Auth0 token: {response.text}")
        
        data = response.json()
        self._token = data["access_token"]
        
        # Set expiry (with 5 minute buffer)
        expires_in = data.get("expires_in", 3600)
        from datetime import timedelta
        self._token_expiry = datetime.now() + timedelta(seconds=expires_in - 300)
        
        return self._token
    
    def create_session(self) -> Optional[str]:
        """Create a new session on the backend and store its ID. Returns session ID."""
        if self.dev_mode:
            self.current_session_id = f"dev-session-{datetime.now().strftime('%Y%m%d%H%M%S')}"
            return self.current_session_id
        try:
            token = self._get_token()
            response = requests.post(
                f"{self.api_url}/sessions",
                headers={"Authorization": f"Bearer {token}", "Content-Type": "application/json"},
                json={},
                timeout=5
            )
            if response.status_code == 201:
                sid = response.json().get("session", {}).get("id")
                self.current_session_id = sid
                return sid
        except Exception as e:
            print(f"Warning: Failed to create session: {e}")
        return None

    def end_session(self) -> None:
        """Close the current session on the backend."""
        if self.dev_mode or not self.current_session_id:
            return
        try:
            token = self._get_token()
            requests.post(
                f"{self.api_url}/sessions/{self.current_session_id}/end",
                headers={"Authorization": f"Bearer {token}", "Content-Type": "application/json"},
                json={},
                timeout=5
            )
        except Exception as e:
            print(f"Warning: Failed to end session: {e}")
        finally:
            self.current_session_id = None

    def execute_action(
        self,
        action_type: str,
        url: str,
        target: Optional[Dict] = None,
        form_data: Optional[Dict] = None,
        domain: Optional[str] = None,
        screenshot: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Execute a browser action through AgentTrust
        
        Args:
            action_type: 'click', 'form_submit', or 'navigation'
            url: Full URL of the page
            target: Target element info (for clicks)
            form_data: Form data (for form submissions)
            domain: Domain (auto-extracted from URL if not provided)
        
        Returns:
            Dict with status, action_id, risk_level, etc.
        """
        if action_type not in ['click', 'form_submit', 'navigation']:
            raise ValueError(f"Invalid action_type: {action_type}")
        
        # Extract domain from URL if not provided
        if not domain:
            from urllib.parse import urlparse
            domain = urlparse(url).netloc
        
        # Dev mode: allow all actions without backend (for browser automation testing)
        if self.dev_mode:
            return {
                "status": "allowed",
                "action_id": f"dev-{datetime.now().strftime('%Y%m%d%H%M%S')}",
                "risk_level": "low",
                "message": "Action allowed (dev mode - no backend)"
            }
        
        # Build action data
        action_data = {
            "type": action_type,
            "url": url,
            "domain": domain,
            "timestamp": datetime.now().isoformat()
        }
        
        if self.current_session_id:
            action_data["sessionId"] = self.current_session_id

        if self.current_prompt_id:
            action_data["promptId"] = self.current_prompt_id
        
        if target:
            action_data["target"] = target
        
        if form_data:
            action_data["form"] = {"fields": form_data}
        
        if screenshot:
            action_data["screenshot"] = screenshot
        
        # Make request
        try:
            token = self._get_token()
        except Exception as e:
            return {
                "status": "error",
                "message": f"Auth failed: {e}. Set AGENTTRUST_DEV_MODE=true to run without backend."
            }
        try:
            response = requests.post(
                f"{self.api_url}/actions",
                headers={
                    "Authorization": f"Bearer {token}",
                    "Content-Type": "application/json"
                },
                json=action_data,
                timeout=10
            )
        except requests.exceptions.RequestException as e:
            return {
                "status": "error",
                "message": f"Backend unreachable: {e}. Is it running? Set AGENTTRUST_DEV_MODE=true to run without backend."
            }
        
        try:
            result = response.json()
        except Exception:
            result = {"error": response.text or "Unknown error"}
        
        if response.status_code == 201:
            return {
                "status": "allowed",
                "action_id": result.get("action", {}).get("id"),
                "risk_level": result.get("action", {}).get("riskLevel"),
                "message": "Action allowed and logged"
            }
        elif response.status_code == 403:
            if result.get("requiresStepUp") and result.get("approvalId"):
                approval_id = result["approvalId"]
                print(f"🔐 High-risk action requires user approval (approvalId={approval_id}). Waiting...")
                approval_result = self.wait_for_approval(approval_id, timeout=60)

                if approval_result.get("approved"):
                    print("✅ Action approved by user. Retrying...")
                    return self._retry_with_approval(action_data, approval_id)
                else:
                    reason = approval_result.get("reason", "User denied the action")
                    print(f"❌ Action denied by user: {reason}")
                    return {
                        "status": "denied",
                        "message": f"Action denied by user: {reason}",
                        "risk_level": result.get("riskLevel"),
                        "approval_denied": True
                    }
            elif result.get("requiresStepUp"):
                return {
                    "status": "step_up_required",
                    "message": "High-risk action requires user approval",
                    "risk_level": result.get("riskLevel"),
                    "error": result.get("error")
                }
            else:
                return {
                    "status": "denied",
                    "message": result.get("error", "Action denied by policy"),
                    "reason": result.get("reason")
                }
        elif response.status_code == 401:
            return {
                "status": "unauthorized",
                "message": "Authentication failed",
                "error": result.get("error")
            }
        else:
            return {
                "status": "error",
                "message": result.get("error", "Unknown error"),
                "status_code": response.status_code
            }
    
    def wait_for_approval(self, approval_id: str, timeout: int = 60) -> Dict[str, Any]:
        """Long-poll the backend waiting for the user to approve or deny an action.
        
        Args:
            approval_id: The approval request ID from a step_up_required response
            timeout: Maximum seconds to wait for user decision
            
        Returns:
            Dict with 'approved' bool and optional 'reason'
        """
        if self.dev_mode:
            return {"approved": False, "reason": "Dev mode - no approval flow"}
        try:
            token = self._get_token()
            response = requests.get(
                f"{self.api_url}/approvals/{approval_id}/wait",
                headers={"Authorization": f"Bearer {token}"},
                params={"timeout": timeout * 1000},
                timeout=timeout + 10
            )
            if response.status_code == 200:
                data = response.json()
                return {
                    "approved": data.get("approved", False),
                    "reason": data.get("reason"),
                    "approvalId": approval_id,
                    "actionId": data.get("actionId")
                }
            elif response.status_code == 404:
                return {"approved": False, "reason": "Approval request not found or expired"}
            else:
                return {"approved": False, "reason": f"Unexpected status: {response.status_code}"}
        except requests.exceptions.Timeout:
            return {"approved": False, "reason": "Approval wait timed out"}
        except Exception as e:
            print(f"⚠️  Approval wait error: {e}")
            return {"approved": False, "reason": str(e)}

    def _retry_with_approval(self, action_data: Dict, approval_id: str) -> Dict[str, Any]:
        """Retry a previously blocked action after receiving user approval."""
        action_data["approvalId"] = approval_id
        try:
            token = self._get_token()
            response = requests.post(
                f"{self.api_url}/actions",
                headers={
                    "Authorization": f"Bearer {token}",
                    "Content-Type": "application/json"
                },
                json=action_data,
                timeout=10
            )
            result = response.json() if response.status_code != 204 else {}
            if response.status_code == 201:
                return {
                    "status": "allowed",
                    "action_id": result.get("action", {}).get("id"),
                    "risk_level": result.get("action", {}).get("riskLevel"),
                    "message": "Action allowed after user approval"
                }
            else:
                return {
                    "status": "error",
                    "message": result.get("error", "Retry after approval failed"),
                    "status_code": response.status_code
                }
        except Exception as e:
            return {"status": "error", "message": f"Retry failed: {e}"}

    def request_step_up(
        self,
        action_data: Dict,
        reason: str
    ) -> Dict[str, Any]:
        """
        Request step-up token for high-risk action
        
        Args:
            action_data: Original action data
            reason: User-provided reason for step-up
        
        Returns:
            Dict with step-up token and expiration
        """
        token = self._get_token()
        response = requests.post(
            f"{self.api_url}/auth/stepup",
            headers={
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json"
            },
            json={
                "action": action_data,
                "reason": reason
            }
        )
        
        result = response.json()
        
        if response.status_code == 200:
            return {
                "success": True,
                "token": result.get("token"),
                "expires_in": result.get("expiresIn"),
                "scopes": result.get("scopes")
            }
        else:
            return {
                "success": False,
                "error": result.get("error", "Step-up failed")
            }
    
    def _update_action_screenshot(self, action_id: str, screenshot: str) -> None:
        """
        Update an action with a screenshot (internal method)
        
        Args:
            action_id: ID of the action to update
            screenshot: Base64 encoded screenshot
        """
        if self.dev_mode:
            return
        try:
            token = self._get_token()
            response = requests.patch(
                f"{self.api_url}/actions/{action_id}",
                headers={
                    "Authorization": f"Bearer {token}",
                    "Content-Type": "application/json"
                },
                json={"screenshot": screenshot}
            )
            if response.status_code != 200:
                detail = ""
                try:
                    detail = f" - {response.json().get('error', response.text[:200])}"
                except Exception:
                    pass
                print(f"⚠️  Failed to update action with screenshot: {response.status_code}{detail}")
        except Exception as e:
            print(f"⚠️  Error updating screenshot: {e}")
    
    def store_prompt(self, content: str, session_id: Optional[str] = None) -> Optional[str]:
        """Store a user prompt and return the prompt ID."""
        if self.dev_mode:
            return None
        try:
            token = self._get_token()
            payload = {"content": content}
            sid = session_id or self.current_session_id
            if sid:
                payload["sessionId"] = sid
            response = requests.post(
                f"{self.api_url}/prompts",
                headers={
                    "Authorization": f"Bearer {token}",
                    "Content-Type": "application/json"
                },
                json=payload,
                timeout=5
            )
            if response.status_code == 201:
                pid = response.json().get("prompt", {}).get("id")
                if pid:
                    self.current_prompt_id = pid
                return pid
        except Exception as e:
            print(f"⚠️  Failed to store prompt: {e}")
        return None

    def update_prompt_response(self, prompt_id: str, response_text: str) -> None:
        """Update a stored prompt with the agent's response."""
        if self.dev_mode or not prompt_id:
            return
        try:
            token = self._get_token()
            requests.patch(
                f"{self.api_url}/prompts/{prompt_id}",
                headers={
                    "Authorization": f"Bearer {token}",
                    "Content-Type": "application/json"
                },
                json={"response": response_text},
                timeout=5
            )
        except Exception as e:
            print(f"⚠️  Failed to update prompt response: {e}")

    def poll_command(self, timeout: int = 30) -> Optional[Dict]:
        """Long-poll the backend for a pending command from the browser extension.
        Returns the command dict if one arrives, or None on timeout."""
        if self.dev_mode or not self.current_session_id:
            return None
        try:
            token = self._get_token()
            response = requests.get(
                f"{self.api_url}/commands/pending",
                headers={"Authorization": f"Bearer {token}"},
                params={"sessionId": self.current_session_id, "timeout": timeout * 1000},
                timeout=timeout + 5
            )
            if response.status_code == 200:
                data = response.json()
                if data.get("success") and data.get("command"):
                    return data["command"]
        except requests.exceptions.Timeout:
            pass
        except (requests.exceptions.ConnectionError, ConnectionResetError, ConnectionError):
            pass
        except Exception as e:
            print(f"⚠️  Command poll error: {e}")
        return None

    def get_audit_log(
        self,
        agent_id: Optional[str] = None,
        domain: Optional[str] = None,
        risk_level: Optional[str] = None,
        start_date: Optional[str] = None,
        end_date: Optional[str] = None,
        limit: int = 100
    ) -> Dict[str, Any]:
        """
        Get audit log
        
        Args:
            agent_id: Filter by agent ID
            domain: Filter by domain
            risk_level: Filter by risk level (low, medium, high)
            start_date: Start date (ISO 8601)
            end_date: End date (ISO 8601)
            limit: Maximum results
        
        Returns:
            Dict with actions array
        """
        if self.dev_mode:
            return {"actions": [], "message": "Dev mode - no audit log"}
        token = self._get_token()
        params = {"limit": limit}
        
        if agent_id:
            params["agentId"] = agent_id
        if domain:
            params["domain"] = domain
        if risk_level:
            params["riskLevel"] = risk_level
        if start_date:
            params["startDate"] = start_date
        if end_date:
            params["endDate"] = end_date
        
        response = requests.get(
            f"{self.api_url}/actions",
            headers={"Authorization": f"Bearer {token}"},
            params=params
        )
        
        return response.json()
    
    def validate_token(self, token: str) -> Dict[str, Any]:
        """
        Validate a JWT token
        
        Args:
            token: JWT token to validate
        
        Returns:
            Validation result
        """
        response = requests.post(
            f"{self.api_url}/auth/validate",
            headers={"Content-Type": "application/json"},
            json={"token": token}
        )
        
        return response.json()


# OpenAI Function definition for ChatGPT
AGENTTRUST_FUNCTION_DEFINITION = {
    "type": "function",
    "function": {
        "name": "agenttrust_browser_action",
        "description": "MANDATORY: Execute a browser action (click, form submit, navigation) through AgentTrust's policy-enforced system. This is the ONLY way to perform browser actions - AgentTrust validation is enforced 100% of the time. Returns whether action is allowed, denied, or requires step-up authentication. You MUST call this function before performing ANY browser automation - there is no alternative or bypass.",
        "parameters": {
            "type": "object",
            "properties": {
                "action_type": {
                    "type": "string",
                    "enum": ["click", "form_submit", "navigation"],
                    "description": "Type of browser action to perform"
                },
                "url": {
                    "type": "string",
                    "description": "Full URL of the page where the action occurs"
                },
                "target": {
                    "type": "object",
                    "description": "Target element information (required for click actions)",
                    "properties": {
                        "tagName": {"type": "string", "description": "HTML tag name"},
                        "id": {"type": "string", "description": "Element ID"},
                        "className": {"type": "string", "description": "CSS class name"},
                        "text": {"type": "string", "description": "Element text content"}
                    }
                },
                "form_data": {
                    "type": "object",
                    "description": "Form data fields (required for form_submit actions)"
                }
            },
            "required": ["action_type", "url"]
        }
    }
}


# Example usage
if __name__ == "__main__":
    # Initialize client
    client = AgentTrustClient()
    
    # Example: Click action
    result = client.execute_action(
        action_type="click",
        url="https://github.com/user/repo",
        target={
            "tagName": "BUTTON",
            "id": "submit-btn",
            "text": "Submit"
        }
    )
    
    print(f"Action status: {result['status']}")
    if result['status'] == 'allowed':
        print(f"Action ID: {result['action_id']}")
        print(f"Risk Level: {result['risk_level']}")
    elif result['status'] == 'step_up_required':
        print("Step-up authentication required")
        # Request step-up
        step_up_result = client.request_step_up(
            action_data={"type": "click", "url": "https://github.com/user/repo"},
            reason="User requested repository deletion after archiving"
        )
        if step_up_result['success']:
            print("Step-up token obtained")
    else:
        print(f"Action denied: {result['message']}")

File: integrations/nemoclaw/src/agenttrust-client.js
const { requestJson, HttpError } = require('./http');

function trimTrailingSlash(value) {
  return String(value || '').replace(/\/+$/, '');
}

function getDomain(url) {
  try {
    return new URL(url).hostname;
  } catch {
    return '';
  }
}

class AgentTrustBridge {
  constructor(options = {}) {
    this.apiBaseUrl = trimTrailingSlash(
      options.apiBaseUrl || process.env.AGENTTRUST_API_URL || 'http://localhost:3000/api'
    );
    this.auth0Domain = trimTrailingSlash(options.auth0Domain || process.env.AUTH0_DOMAIN);
    this.auth0ClientId = options.auth0ClientId || process.env.AUTH0_CLIENT_ID;
    this.auth0ClientSecret = options.auth0ClientSecret || process.env.AUTH0_CLIENT_SECRET;
    this.auth0Audience = trimTrailingSlash(options.auth0Audience || process.env.AUTH0_AUDIENCE);
    this.userEmail = options.userEmail || process.env.AGENTTRUST_USER_EMAIL || '';
    this.userPassword = options.userPassword || process.env.AGENTTRUST_USER_PASSWORD || '';

    this.agentToken = null;
    this.agentTokenExpiresAt = 0;
    this.userToken = options.userToken || null;
    this.currentSessionId = null;
    this.currentPromptId = null;
  }

  async getAgentToken(forceRefresh = false) {
    if (!forceRefresh && this.agentToken && Date.now() < this.agentTokenExpiresAt) {
      return this.agentToken;
    }

    if (!this.auth0Domain || !this.auth0ClientId || !this.auth0ClientSecret || !this.auth0Audience) {
      throw new Error(
        'Missing Auth0 M2M configuration. Set AUTH0_DOMAIN, AUTH0_CLIENT_ID, AUTH0_CLIENT_SECRET, and AUTH0_AUDIENCE.'
      );
    }

    const data = await requestJson(`https://${this.auth0Domain}/oauth/token`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        client_id: this.auth0ClientId,
        client_secret: this.auth0ClientSecret,
        audience: this.auth0Audience,
        grant_type: 'client_credentials',
      }),
    });

    this.agentToken = data.access_token;
    this.agentTokenExpiresAt = Date.now() + Math.max((data.expires_in || 3600) - 300, 60) * 1000;
    return this.agentToken;
  }

  async loginUser({ email, password } = {}) {
    const resolvedEmail = email || this.userEmail;
    const resolvedPassword = password || this.userPassword;

    if (!resolvedEmail || !resolvedPassword) {
      throw new Error('User login requires an email and password.');
    }

    const data = await requestJson(`${this.apiBaseUrl}/users/login`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        email: resolvedEmail,
        password: resolvedPassword,
      }),
    });

    this.userToken = data.token;
    return data;
  }

  async ensureUserToken() {
    if (this.userToken) {
      return this.userToken;
    }

    await this.loginUser();
    return this.userToken;
  }

  async requestAgent(pathname, options = {}) {
    const token = await this.getAgentToken();
    return requestJson(`${this.apiBaseUrl}${pathname}`, {
      ...options,
      headers: {
        ...(options.headers || {}),
        Authorization: `Bearer ${token}`,
      },
    });
  }

  async requestUser(pathname, options = {}) {
    const token = await this.ensureUserToken();
    return requestJson(`${this.apiBaseUrl}${pathname}`, {
      ...options,
      headers: {
        ...(options.headers || {}),
        Authorization: `Bearer ${token}`,
      },
    });
  }

  async verifyConnectivity() {
    const healthUrl = this.apiBaseUrl.replace(/\/api$/, '/health');
    await requestJson(healthUrl);
    await this.getAgentToken();
    return { ok: true };
  }

  async createSession() {
    const data = await this.requestAgent('/sessions', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({}),
    });

    this.currentSessionId = data?.session?.id || null;
    if (this.currentSessionId && this.userToken) {
      await this.claimSession(this.currentSessionId);
    }
    return data.session;
  }

  async endSession(sessionId = this.currentSessionId) {
    if (!sessionId) return null;

    const data = await this.requestAgent(`/sessions/${sessionId}/end`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({}),
    });

    if (this.currentSessionId === sessionId) {
      this.currentSessionId = null;
    }
    return data.session;
  }

  async storePrompt(content, sessionId = this.currentSessionId) {
    const data = await this.requestAgent('/prompts', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        content,
        sessionId: sessionId || undefined,
      }),
    });

    this.currentPromptId = data?.prompt?.id || null;
    return data.prompt;
  }

  async updatePromptProgress(promptId, progress) {
    if (!promptId) return null;
    return this.requestAgent(`/prompts/${promptId}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ progress }),
    });
  }

  async updatePromptResponse(promptId, response) {
    if (!promptId) return null;
    return this.requestAgent(`/prompts/${promptId}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ response }),
    });
  }

  async uploadScreenshot(actionId, screenshot) {
    if (!actionId || !screenshot) return null;
    return this.requestAgent(`/actions/${actionId}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ screenshot }),
    });
  }

  async executeAction(actionData, options = {}) {
    const payload = {
      ...actionData,
      domain: actionData.domain || getDomain(actionData.url),
      sessionId: actionData.sessionId || this.currentSessionId || undefined,
      promptId: actionData.promptId || this.currentPromptId || undefined,
    };

    try {
      const data = await this.requestAgent('/actions', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload),
      });

      return {
        status: 'allowed',
        action_id: data?.action?.id,
        risk_level: data?.action?.riskLevel,
        executionLease: data?.executionLease || null,
        message: 'Action allowed and logged',
        raw: data,
      };
    } catch (error) {
      if (!(error instanceof HttpError) || error.status !== 403) {
        throw error;
      }

      const data = error.data || {};
      if (!data.requiresStepUp || !data.approvalId) {
        return {
          status: 'denied',
          message: data.error || error.message,
          risk_level: data.riskLevel,
          raw: data,
        };
      }

      if (options.autoWaitForApproval === false) {
        return {
          status: 'step_up_required',
          message: data.error || 'Action requires approval',
          risk_level: data.riskLevel,
          approvalId: data.approvalId,
          action_id: data.actionId,
          raw: data,
        };
      }

      const approvalResult = await this.waitForApproval(data.approvalId, options.approvalTimeoutSeconds || 120);
      if (!approvalResult.approved) {
        return {
          status: 'denied',
          message: approvalResult.reason || 'Action denied or approval timed out',
          risk_level: data.riskLevel,
          approval_denied: true,
          approvalId: data.approvalId,
          raw: data,
        };
      }

      return this.retryActionWithApproval(payload, data.approvalId);
    }
  }

  async retryActionWithApproval(actionData, approvalId) {
    try {
      const data = await this.requestAgent('/actions', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          ...actionData,
          approvalId,
        }),
      });

      return {
        status: 'allowed',
        action_id: data?.action?.id,
        risk_level: data?.action?.riskLevel,
        executionLease: data?.executionLease || null,
        message: 'Action allowed after user approval',
        raw: data,
      };
    } catch (error) {
      if (error instanceof HttpError) {
        return {
          status: 'error',
          message: error.message,
          status_code: error.status,
          raw: error.data,
        };
      }
      throw error;
    }
  }

  async waitForApproval(approvalId, timeoutSeconds = 120) {
    try {
      const data = await this.requestAgent(`/approvals/${approvalId}/wait?timeout=${Math.min(timeoutSeconds, 60) * 1000}`, {
        method: 'GET',
      });

      return {
        approved: !!data.approved,
        reason: data.reason,
        actionId: data.actionId,
        approvalId,
      };
    } catch (error) {
      if (error instanceof HttpError && error.status === 404) {
        return { approved: false, reason: 'Approval not found or expired', approvalId };
      }
      throw error;
    }
  }

  async respondToApproval(approvalId, approved) {
    return this.requestUser(`/approvals/${approvalId}/respond`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ approved: !!approved }),
    });
  }

  async listPendingApprovals(sessionId) {
    const query = sessionId ? `?sessionId=${encodeURIComponent(sessionId)}` : '';
    const data = await this.requestUser(`/approvals/pending${query}`, { method: 'GET' });
    return data.approvals || [];
  }

  async sendCommand(content, sessionId = this.currentSessionId) {
    if (!sessionId) {
      throw new Error('A sessionId is required to send a command.');
    }

    return this.requestUser('/commands', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ content, sessionId }),
    });
  }

  async claimSession(sessionId = this.currentSessionId) {
    if (!sessionId) {
      throw new Error('A sessionId is required to claim a session.');
    }

    const data = await this.requestUser(`/sessions/${sessionId}/claim`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({}),
    });

    return data.session;
  }

  async pollCommand(sessionId = this.currentSessionId, timeoutSeconds = 30) {
    if (!sessionId) return null;

    const data = await this.requestAgent(
      `/commands/pending?sessionId=${encodeURIComponent(sessionId)}&timeout=${Math.min(timeoutSeconds, 30) * 1000}`,
      { method: 'GET' }
    );
    return data.command || null;
  }

  async getSession(sessionId = this.currentSessionId) {
    if (!sessionId) {
      throw new Error('A sessionId is required to load session details.');
    }

    const data = await this.requestUser(`/sessions/${sessionId}`, { method: 'GET' });
    return data.session;
  }

  async listSessions({ agentId = 'all', limit = 20, fields = 'full' } = {}) {
    const params = new URLSearchParams({
      agentId,
      limit: String(limit),
      fields,
    });
    const data = await this.requestUser(`/sessions?${params.toString()}`, { method: 'GET' });
    return data.sessions || [];
  }

  async callExternalApi({ provider, method, url, body, sessionId, promptId, approvalId }, options = {}) {
    const payload = {
      provider,
      method,
      url,
      body,
      sessionId: sessionId || this.currentSessionId || undefined,
      promptId: promptId || this.currentPromptId || undefined,
      userToken: this.userToken || undefined,
      issueLeaseOnly: !!options.issueLeaseOnly,
      approvalId: approvalId || options.approvalId || undefined,
    };

    try {
      return await this.requestAgent('/external/call', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload),
      });
    } catch (error) {
      if (!(error instanceof HttpError) || error.status !== 403) {
        throw error;
      }

      const data = error.data || {};
      if (!data.requiresStepUp || !data.approvalId) {
        return {
          success: false,
          status: 'denied',
          error: data.error || error.message,
          riskLevel: data.riskLevel,
        };
      }

      if (options.autoWaitForApproval === false) {
        return {
          success: false,
          status: 'step_up_required',
          error: data.error || 'API call requires approval',
          approvalId: data.approvalId,
          riskLevel: data.riskLevel,
        };
      }

      const approvalResult = await this.waitForApproval(data.approvalId, options.approvalTimeoutSeconds || 120);
      if (!approvalResult.approved) {
        return {
          success: false,
          status: 'denied',
          error: approvalResult.reason || 'API call denied or approval timed out',
          approvalId: data.approvalId,
          riskLevel: data.riskLevel,
        };
      }

      return this.requestAgent('/external/call', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          ...payload,
          approvalId: data.approvalId,
        }),
      });
    }
  }
}

module.exports = {
  AgentTrustBridge,
  getDomain,
};
